aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -386,51 +386,108 @@ inline Tensor reduce_multi_associative(const Tensor &self, IntList dims_, bool k
     return self;
   }
   int64_t ndims = self.dim();
-  auto reduce_dims = dim_list_to_bitset(dims_, ndims);
-  Tensor result = self;
-  for (int64_t dim = ndims-1; dim >= 0; dim--) {
-    if (reduce_dims[dim])
-      result = reduce_1(result, dim, keepdim);
+  auto reduced_size = self.sizes().vec();
+  auto dims = dims_.vec();
+  maybe_wrap_dims(dims, ndims);
+  // Sort the reduced dimensions so that we reduce the largest dimension first.
+  std::sort(dims.begin(), dims.end(),
+        [&](int64_t i, int64_t j){ return reduced_size[i] > reduced_size[j]; });
+  int64_t reduced_numel = self.numel();
+  int64_t max_reduced_numel = reduced_numel / reduced_size[dims[0]];
+  int64_t buffer_size = max_reduced_numel + max_reduced_numel / reduced_size[dims[1]];
```

Comment:
The use of the variable name `reduced_size` here is confusing, because the size isn't reduced yet (it will be in the loop, but none of that has occurred yet.)